[lld][WebAssembly] Fix non-pie dynamic-linking executable

yamt opened the pull request:
The commit 22b7b84860d39da71964c9b329937f2ee1d875ba
made the symbols provided by shared libraries "defined",
and thus effectively made it impossible to generate non-pie
dynamically linked executables using --unresolved-symbols=import-dynamic.

This commit, based on https://github.com/llvm/llvm-project/pull/109249,
fixes it by checking sym->isShared() explictly.
(as a bonus, you don't need to rely on --unresolved-symbols=import-dynamic
anymore.)

Fixes https://github.com/llvm/llvm-project/issues/107387

sbc100 commented on lld/wasm/Driver.cpp:
```diff
@@ -913,7 +913,8 @@ static void createSyntheticSymbols() {
   if (ctx.isPic ||
-      config->unresolvedSymbols == UnresolvedPolicy::ImportDynamic) {
+      config->unresolvedSymbols == UnresolvedPolicy::ImportDynamic ||
+      !config->isStatic) {
```
I wonder if can replace `ctx.isPic` with ` !config->isStatic` above?    Does it make sense to allow `-shared` + `-static`?  Or `-pie` + `-static`?

I wonder if we need a better way to detect if dynamic linking used at all since you can imagine a case where `isStatic` is true, but we do have dynamic linking.  For example:

```
$ wasm-ld foo.o -Bdynamic -ldylib -Bstatic -lstaticlib
```

In that case since `-Bstatic` comes last on the command line the `config->isStatic` will be false even though we include libdylib.so dyanmically.

Another example:

```
wasm-ld foo.o libdylib.so libstaticlib.a
```

Here `config->isStatic` will remain false even though dynamic linking is used.
```

sbc100 replied:
Actually it looks like that the latter case should reject the `.so` file.  I see the ELF linker does this: https://github.com/llvm/llvm-project/blob/96b7c64b8a874584a9dad44bb8901904c14701c0/lld/ELF/Driver.cpp#L341-L344.  I'll upload a PR for that.

yamt replied:
> I wonder if can replace `ctx.isPic` with ` !config->isStatic` above? Does it make sense to allow `-shared` + `-static`? Or `-pie` + `-static`?
> 
> I wonder if we need a better way to detect if dynamic linking used at all since you can imagine a case where `isStatic` is true, but we do have dynamic linking. For example:
> 
> ```
> $ wasm-ld foo.o -Bdynamic -ldylib -Bstatic -lstaticlib
> ```
> 
> In that case since `-Bstatic` comes last on the command line the `config->isStatic` will be false even though we include libdylib.so dyanmically.

i guess it's simpler to defer the decision to createApplyDataRelocationsFunction.
how do you think?

yamt replied:
> i guess it's simpler to defer the decision to createApplyDataRelocationsFunction.

i pushed a commit to implement it.

yamt replied:
@sbc100 do you have any more concerns?

sbc100 commented:
Could you update the PR description describing what was broken and how this fixes the issue?

Also, can we add a new test or update one of the existing ones to make sure we don't regress again?

yamt committed:
[lld][WebAssembly] Fix non-pie dynamic-linking executable

The commit 22b7b84860d39da71964c9b329937f2ee1d875ba
made the symbols provided by shared libraries "defined",
and thus effectively made it impossible to generate non-pie
dynamically linked executables using --unresolved-symbols=import-dynamic.

This commit, based on https://github.com/llvm/llvm-project/pull/109249,
fixes it by checking sym->isShared() explictly.
(as a bonus, you don't need to rely on --unresolved-symbols=import-dynamic
anymore.)

Fixes https://github.com/llvm/llvm-project/issues/107387

yamt commented on lld/wasm/Writer.cpp:
```diff
@@ -1467,15 +1469,29 @@ void Writer::createApplyDataRelocationsFunction() {
+      make<SyntheticFunction>(nullSignature, "__wasm_apply_data_relocs"));
+  def->markLive();
+
+  createFunction(def, bodyContent);
```
> This change seems to incorporate a couple of different things now. Would you mind separating out the change to "Only create __wasm_apply_data_relocs conditionally". That seems like it could land separately and is somewhat unrelated to fix at hand?

https://github.com/llvm/llvm-project/pull/109249

> Also, regarding the conditional creation of this symbol, what happens if somebody does `--export=__wasm_apply_data_relocs` or otherwise references this symbol, but we "skip" creating it?

is it expected to work at all? do you have real use cases? or is it just a theoretical concern?
i thought __wasm_apply_data_relocs was just for dynamic-linking.

yamt commented:
> Could you update the PR description describing what was broken and how this fixes the issue?
> 
> Also, can we add a new test or update one of the existing ones to make sure we don't regress again?

i will update if/when https://github.com/llvm/llvm-project/pull/109249 is merged.
i will mark this PR a draft for now.

yamt committed:
add a test

yamt commented:
* updated after https://github.com/llvm/llvm-project/pull/109249
* added a test

yamt commented on lld/test/wasm/dylink-non-pie.s:
```diff
@@ -0,0 +1,35 @@
+# RUN: llvm-mc -filetype=obj -triple=wasm32-unknown-unknown -o %t.lib.o %p/Inputs/lib.s
+# RUN: wasm-ld -m wasm32 --experimental-pic -shared --no-entry %t.lib.o -o %t.lib.so
+# RUN: llvm-mc -filetype=obj -mattr=+reference-types -triple=wasm32-unknown-unknown -o %t.o %s
+# RUN: wasm-ld -m wasm32 --features=mutable-globals -Bdynamic --export-table --growable-table --export-memory --export=__stack_pointer --export=__heap_base --export=__heap_end --entry=_start %t.o %t.lib.so -o %t.wasm
```
i wanted to mimic the real situation to link a non-pie executable.

yamt replied:
done

yamt commented on lld/test/wasm/Inputs/lib.s:
```diff
@@ -0,0 +1,6 @@
+	.type	f,@function
+f:
+	.functype	f () -> ()
+	end_function
```
ok. i will try.

yamt commented on lld/wasm/Relocations.cpp:
```diff
@@ -146,7 +146,8 @@ void scanRelocations(InputChunk *chunk) {
         (sym->isUndefined() &&
-         config->unresolvedSymbols == UnresolvedPolicy::ImportDynamic)) {
+         config->unresolvedSymbols == UnresolvedPolicy::ImportDynamic) ||
+        sym->isShared()) {
```
ok

yamt replied:
done

yamt commented on lld/test/wasm/dylink-non-pie.s:
```diff
@@ -0,0 +1,35 @@
+	i32.const	f_p@MBREL
+	i32.add
+	i32.load	0
+	call_indirect	 __indirect_function_table, () -> ()
```
let me think a bit. thank you.

yamt replied:
done

yamt committed:
dylink-non-pie.s: use the existing file

yamt committed:
tweak the order of conditions

no functional changes are intended

yamt committed:
lld/test/wasm/dylink-non-pie.s: simplify

sbc100 commented on lld/test/wasm/dylink-non-pie.s:
```diff
@@ -0,0 +1,29 @@
+# RUN: llvm-mc -filetype=obj -triple=wasm32-unknown-unknown -o %t.lib.o %p/Inputs/ret32.s
+# RUN: wasm-ld -m wasm32 --experimental-pic -shared --no-entry %t.lib.o -o %t.lib.so
+# RUN: llvm-mc -filetype=obj -triple=wasm32-unknown-unknown -o %t.o %s
+# RUN: wasm-ld -m wasm32 -Bdynamic %t.o %t.lib.so -o %t.wasm
```
Just to confirm, this effectively generates a non-PIC output?   And `ctx.isPic` not set?

I wonder if we can confirm this this test somehow?   Perhaps its enough that we successfully use `i32.const  f_p` (which would not work if the output was PIC, right?) instead of `i32.const   f_p@MBREL`

yamt replied:
`ctx.isPic` is not set as neither of pie or shared is set.

the following is the generated output. it looks reasonable to me.
```
(module $dylink-non-pie.s.tmp.wasm
  (type (;0;) (func (param f32) (result i32)))
  (type (;1;) (func))
  (import "env" "ret32" (func $ret32 (type 0)))
  (import "GOT.func" "ret32" (global $ret32 (mut i32)))
  (func $__wasm_apply_data_relocs (type 1)
    i32.const 1024
    global.get $ret32
    i32.store)
  (func $_start (type 1)
    i32.const 1024
    drop)
  (table (;0;) 1 1 funcref)
  (memory (;0;) 2)
  (global $__stack_pointer (mut i32) (i32.const 66576))
  (export "memory" (memory 0))
  (export "_start" (func $_start))
  (export "__wasm_apply_data_relocs" (func $__wasm_apply_data_relocs))
  (data $.data (i32.const 1024) "\00\00\00\00"))
```

yamt replied:
> I wonder if we can confirm this this test somehow? Perhaps its enough that we successfully use i32.const  f_p (which would not work if the output was PIC, right?) instead of i32.const   f_p@MBREL

maybe we can confirm it doesn't have __memory_base/__table_base imports?

yamt replied:
done

yamt commented on lld/test/wasm/dylink-non-pie.s:
```diff
@@ -0,0 +1,29 @@
+
+# CHECK:   - Type:            EXPORT
+# CHECK:       - Name:            __wasm_apply_data_relocs
+# CHECK-NEXT:         Kind:            FUNCTION
```
do you know any existing tests doing similar things which i can use as an example?

sbc100 replied:
Yes, there are a few tests that disassembly stuff.  Looks for `DIS-NEXT` in lld/test/wasm.

yamt replied:
ok. i will take a look.

yamt replied:
done

yamt committed:
lld/test/wasm/dylink-non-pie.s: check the contents of __wasm_apply_data_relocs

yamt committed:
lld/test/wasm/dylink-non-pie.s: ensure it doesn't import __memory_base

yamt commented:
@sbc100 do you have any more concerns?

sbc100 commented on lld/test/wasm/dylink-non-pie.s:
```diff
@@ -0,0 +1,41 @@
+	.functype	ret32 (f32) -> (i32)
+	.functype	_start () -> ()
+	.globl	_start
+	.type	_start,@function
```
this line is not needed I think

yamt replied:
removed

sbc100 commented on lld/test/wasm/dylink-non-pie.s:
```diff
@@ -0,0 +1,41 @@
+# RUN: llvm-objdump -d --no-show-raw-insn --no-leading-addr %t.wasm | FileCheck %s --check-prefixes DIS
+
+	.functype	ret32 (f32) -> (i32)
+	.functype	_start () -> ()
```
This line is not needed since its already on line 13

yamt replied:
removed

sbc100 reviewed (approved):
lgtm % nits

yamt committed:
lld/test/wasm/dylink-non-pie.s: remove a few unnecessary things

yamt commented on lld/test/wasm/dylink-non-pie.s:
```diff
@@ -0,0 +1,41 @@
+	end_function
+
+	.section	.data.f_p,"",@
+	.globl	f_p
```
it doesn't seem necessary. removed. thank you.

yamt commented:
can this land?
